### PR TITLE
Add test case to verify filters in RHODS Dashboard resources page  (ODS-489)

### PR DESCRIPTION
**Pull request by rkamble5**
Signed-off-by: rkamble5 <rkamble@redhat.com>

### CODE REVIEW

File: tests/Tests/400__ods_dashboard/400__ods_dashboard.robot [hunk 1]
```diff
@@ -113,6 +113,50 @@ Search and Verify GPU Items Appears In Resources Page
     Search Items In Resources Section   GPU
     Check GPU Resources
 
+Verify Advanced filtering by status, provider, types from Resources tab is working as expected
+   [Documentation]    check if it is possible to filter items by enabling various filters like status,provider
+   [Tags]    ODS-489
```
Comment (milindw96) [inline tests/Tests/400__ods_dashboard/400__ods_dashboard.robot:118]: could u plz add suitable tags here 
example:
sanity or tier*

**Review by milindw96 — CHANGES_REQUESTED** [review]

File: tests/Tests/400__ods_dashboard/400__ods_dashboard.robot [hunk 2]
```diff
@@ -113,6 +113,52 @@ Search and Verify GPU Items Appears In Resources Page
     Search Items In Resources Section   GPU
     Check GPU Resources
 
+Verify Advanced filtering by status, provider, types from Resources tab is working as expected
```
Comment (jgarciao) [inline tests/Tests/400__ods_dashboard/400__ods_dashboard.robot:116]: Test case title should be in TitleCase

Main tests should be simplified in simpler keywords. See the style guide for test complexity examples

**Comment by rkamble5** [comment]
Tested successfully here :- https://opendatascience-jenkins-csb-rhods.apps.ocp-c1.prod.psi.redhat.com/job/rhods-ci-pr-test/597/

File: tests/Tests/400__ods_dashboard/400__ods_dashboard.robot [hunk 3]
```diff
@@ -160,13 +207,33 @@ Check GPU Resources
...
+
+Deselct Checkbox Using Id
+    [Documentation]    Deselect check-box
+    [Arguments]    ${id}
+    Click Element    id=${id}
+    sleep    10s
+
+Verify The Cards Using Selector And It Value
+    [Documentation]    virifies the items
+    [Arguments]    ${selector}    ${list_of_items}    ${index_of_text}=0
```
Comment (bdattoma) [inline tests/Tests/400__ods_dashboard/400__ods_dashboard.robot:234]: could you find a more self-explanatory name for "index_of_text" ?

File: tests/Tests/400__ods_dashboard/400__ods_dashboard.robot [hunk 4]
```diff
@@ -98,28 +99,74 @@ Verify Documentation Link HTTP Status Code
...
+    ...       Tier1
+    Click Link    Resources
+    Wait Until Page Contains    Self-managed
+    #Filter by enabled filter
+    Select Checkbox Using Id    enabled-filter-checkbox--check-box
+    @{expected_items}=    Create List    Creating a Jupyter notebook
+    ...    Deploying a sample Python application using Flask and OpenShift
+    ...    How to install Python packages on your notebook server    How to update notebook server settings
+    ...    How to use data from Amazon S3 buckets    How to view installed packages on your notebook server
+    ...    JupyterHub
```
Comment (bdattoma) [inline tests/Tests/400__ods_dashboard/400__ods_dashboard.robot:129]: maybe you could wrap these lines in a kw like `Filter Resources by Enabled`

File: tests/Tests/400__ods_dashboard/400__ods_dashboard.robot [hunk 5]
```diff
@@ -98,28 +99,74 @@ Verify Documentation Link HTTP Status Code
...
+    Verify The Cards Using Selector And It Value    selector=pf-c-card__title odh-card__doc-title
+    ...    list_of_items=${expected_items}
```
Comment (bdattoma) [inline tests/Tests/400__ods_dashboard/400__ods_dashboard.robot:131]: Could you find a more self-explanatory name for this kw?

File: tests/Tests/400__ods_dashboard/400__ods_dashboard.robot (hunk 4 above)
Comment (bdattoma) [inline tests/Tests/400__ods_dashboard/400__ods_dashboard.robot:129]: maybe you could wrap these lines in a kw like Filter Resources by Status "Enabled"

The same is valid for the other filters you've applied in this test case

File: tests/Tests/400__ods_dashboard/400__ods_dashboard.robot [hunk 6]
```diff
@@ -98,28 +99,74 @@ Verify Documentation Link HTTP Status Code
...
+    Search Items In Resources Section    GPU
     Check GPU Resources
 
+Verify filters Are Working On Resources Page
+    [Documentation]    check if it is possible to filter items by enabling various filters like status,provider
+    [Tags]    Sanity
+    ...       ODS-489
+    ...       Tier1
+    Click Link    Resources
+    Wait Until Page Contains    Self-managed
```
Comment (bdattoma) [inline tests/Tests/400__ods_dashboard/400__ods_dashboard.robot:122]: create a kw like `Wait Until Resource Page Is Loaded`

File: tests/Tests/400__ods_dashboard/400__ods_dashboard.robot [hunk 7]
```diff
@@ -160,13 +207,33 @@ Check GPU Resources
...
-        Page Should Not Contain     //a[@href="https://github.com/ContinuumIO/gtc2018-numba"]
+        Page Should Not Contain Element     //article[@id="python-gpu-numba-tutorial"]
+        Page Should Not Contain Element     //a[@href="https://github.com/ContinuumIO/gtc2018-numba"]
     END
 
+Select Checkbox Using Id
+    [Documentation]    Select check-box
+    [Arguments]    ${id}
+    Select Checkbox    id=${id}
+    sleep    10s
```
Comment (bdattoma) [inline tests/Tests/400__ods_dashboard/400__ods_dashboard.robot:224]: would you try to replace the 10 seconds of sleep with a wait mechanism?

File: tests/Tests/400__ods_dashboard/400__ods_dashboard.robot [hunk 8]
```diff
@@ -160,13 +207,33 @@ Check GPU Resources
...
+
+Deselct Checkbox Using Id
+    [Documentation]    Deselect check-box
+    [Arguments]    ${id}
+    Click Element    id=${id}
+    sleep    10s
```
Comment (bdattoma) [inline tests/Tests/400__ods_dashboard/400__ods_dashboard.robot:230]: same as above

File: tests/Tests/400__ods_dashboard/400__ods_dashboard.robot [hunk 9]
```diff
@@ -160,13 +207,33 @@ Check GPU Resources
...
+Select Checkbox Using Id
+    [Documentation]    Select check-box
+    [Arguments]    ${id}
+    Select Checkbox    id=${id}
+    sleep    10s
+
+Deselct Checkbox Using Id
+    [Documentation]    Deselect check-box
+    [Arguments]    ${id}
+    Click Element    id=${id}
```
Comment (bdattoma) [inline tests/Tests/400__ods_dashboard/400__ods_dashboard.robot:229]: why not using `[Unselect Checkbox](https://robotframework.org/SeleniumLibrary/SeleniumLibrary.html#Unselect%20Checkbox)` to be consistent with `Select Checkbox Using Id`?

File: tests/Tests/400__ods_dashboard/400__ods_dashboard.robot [hunk 10]
```diff
@@ -98,28 +99,74 @@ Verify Documentation Link HTTP Status Code
...
+    ...    How to use data from Amazon S3 buckets    How to view installed packages on your notebook server
+    ...    JupyterHub
+    Verify The Cards Using Selector And It Value    selector=pf-c-card__title odh-card__doc-title
+    ...    list_of_items=${expected_items}
+    Deselct Checkbox Using Id    enabled-filter-checkbox--check-box
+    # Filter by application (aka provider)
+    @{expected_items}=    Create List    by Anaconda Commercial Edition
+    Select Checkbox Using Id    Anaconda Commercial Edition--check-box
+    Verify The Cards Using Selector And It Value    selector=pf-c-card__title odh-card__doc-title
+    ...    list_of_items=${expected_items}    index_of_text=1
```
Comment (bdattoma) [inline tests/Tests/400__ods_dashboard/400__ods_dashboard.robot:137]: maybe you can define this list (and all the others) in the variables section of RF

File: tests/Tests/400__ods_dashboard/400__ods_dashboard.robot (hunk 3 above)
Comment (rkamble5) [reply tests/Tests/400__ods_dashboard/400__ods_dashboard.robot:234]: done I have provided documentation

File: tests/Tests/400__ods_dashboard/400__ods_dashboard.robot (hunk 10 above)
Comment (rkamble5) [reply tests/Tests/400__ods_dashboard/400__ods_dashboard.robot:137]: Is ok if I do not change the declaration of lists
maybe if we declare in the variable section there will be too many variable with different names

File: tests/Tests/400__ods_dashboard/400__ods_dashboard.robot [hunk 11]
```diff
@@ -98,28 +99,44 @@ Verify Documentation Link HTTP Status Code
...
 
+Verify filters Are Working On Resources Page
+    [Documentation]    check if it is possible to filter items by enabling various filters like status,provider
+    [Tags]    Sanity
+    ...       ODS-489
+    ...       Tier1
+    Click Link    Resources
+    Wait Until Resource Page Is Loaded
+    Filter Resources By Status "Enabled"
+    Filter By Resource Type
```
Comment (bdattoma) [inline tests/Tests/400__ods_dashboard/400__ods_dashboard.robot:124]: hmm to make it more clear, would you call it "Filter By Resource Type And Check Output"? so it is clear that you're checking the result

File: tests/Tests/400__ods_dashboard/400__ods_dashboard.robot [hunk 12]
```diff
@@ -160,13 +177,89 @@ Check GPU Resources
...
+    sleep    10s
+
+Deselect Checkbox Using Id
+    [Documentation]    Deselect check-box
+    [Arguments]    ${id}
+    Unselect Checkbox    id=${id}
+    sleep    10s
+
+Verify The Resources Are Filtered
+    [Documentation]    virifies the items, ${index_of_text} is index text appear on resource 0 = title,1=provider,2=tag(like documentation,tutorial)
```
Comment (bdattoma) [inline tests/Tests/400__ods_dashboard/400__ods_dashboard.robot:203]: hm maybe "verified the items" is too generic.

File: tests/Tests/400__ods_dashboard/400__ods_dashboard.robot (hunk 10 above)
Comment (bdattoma) [reply tests/Tests/400__ods_dashboard/400__ods_dashboard.robot]: I think it wouldn't be a big problem to have 5 variables in the variable sections

File: tests/Tests/400__ods_dashboard/400__ods_dashboard.robot [hunk 13]
```diff
@@ -98,28 +115,45 @@ Verify Documentation Link HTTP Status Code
...
+    [Tags]    Sanity
+    ...       ODS-489
+    ...       Tier1
+    Click Link    Resources
+    Wait Until Resource Page Is Loaded
+    Filter Resources By Status "Enabled" And Check Output
+    Filter By Resource Type And Check Output
+    Filter By Provider Type And Check Output
+    Filter By Application (Aka Povider) And Check Output
+    @{list_of_ids_for_combinations}    Create List    documentation--check-box    Red Hat managed--check-box
```
Comment (bdattoma) [inline tests/Tests/400__ods_dashboard/400__ods_dashboard.robot:144]: could you put this to variable section as well?

**Comment by rkamble5** [comment]
Tested successfully here :- https://opendatascience-jenkins-csb-rhods.apps.ocp-c1.prod.psi.redhat.com/job/rhods-ci-pr-test/679/

**Review by jgarciao — DISMISSED** [review]
lgtm

File: tests/Tests/400__ods_dashboard/400__ods_dashboard.robot [hunk 14]
```diff
@@ -100,28 +115,41 @@ Verify Documentation Link HTTP Status Code
...
+    Verify Username Displayed On RHODS Dashboard    ${TEST_USER.USERNAME}
 
 Search and Verify GPU Items Appears In Resources Page
-    [Tags]  ODS-1226
-    Search Items In Resources Section   GPU
+    [Tags]    ODS-1226
+    Search Items In Resources Section    GPU
     Check GPU Resources
 
+Verify filters Are Working On Resources Page
```
[the remark this answers is not in the document]
Comment (rkamble5) [reply tests/Tests/400__ods_dashboard/400__ods_dashboard.robot]: done

**Comment by tarukumar** [comment]
LGTM

**Review by tarukumar — APPROVED** [review]

**Review by mwaykole — APPROVED** [review]

**Pull request merged**